### PR TITLE
Use go-based mock-smf

### PR DESCRIPTION
**Pull request by EmanueleGallone**
Update STCs to use go based mock-smf (pfcpsim)

### CODE REVIEW

**Comment by EmanueleGallone** [comment]
test this please

**Comment by EmanueleGallone** [comment]
retest this please

**Comment by EmanueleGallone** [comment]
retest this please

**Comment by EmanueleGallone** [comment]
retest this please

**Comment by EmanueleGallone** [comment]
retest this please

**Comment by EmanueleGallone** [comment]
test this please

File: scenarios/bin/onos-cmd [hunk 1]
```diff
@@ -2,12 +2,7 @@
...
-fi
-
-# Use sshpass to skip the password prompt
-docker run -it --rm --network host ictu/sshpass \
-  -procks ssh -o "UserKnownHostsFile=/dev/null" \
-  -o "StrictHostKeyChecking=no" -o LogLevel=ERROR -p 8101 onos@localhost "$@"
+node=$1
+cliCmd=$2
+output=$3
+onos "$node" "$cliCmd" >> "$output"
```
Comment (daniele-moro) [inline scenarios/bin/onos-cmd:8]: any reason to change this? This would affect the `set-app-info` and `set-app-debug` make target in the `scenarios/Makefile`

File: scenarios/pfcp-agent-failure.xml [hunk 2]
```diff
@@ -7,14 +7,17 @@
...
             <step name="AF-Modify-Session" requires="^" delay="1"
-                  exec="mock-smf-cmd modify"/>
+                  exec="${DOCKER_COMPOSE_CMD} exec -T mock-smf pfcpctl -c modify --ue-pool 17.0.0.0/24 --gnb-addr 140.0.100.1"/>
         </group>
 
         <step name="AF-Check-Flows-1" requires="AF-Setup-Fwd" delay="2"
               exec="onos-cli-grep ${OCI} up4:read-flows 'UL flows=1, DL flows=1'"/>
 
+        <step name="AF-Pre-Check-Interfaces" requires="^"
+              exec="onos-cmd ${OCI} 'up4:read-entities -i -g' 'tmp/Pre_check_interfaces.log'"/>
```
Comment (daniele-moro) [inline scenarios/pfcp-agent-failure.xml:19]: Remove

File: scenarios/pfcp-agent-failure.xml [hunk 3]
```diff
@@ -24,14 +27,25 @@
         <step name="AF-Check-Flows-2" requires="^"
               exec="onos-cli-grep ${OCI} up4:read-flows 'UL flows=0, DL flows=0'"/>
 
+        <step name="AF-Mid-Check-Interfaces" requires="^"
+              exec="onos-cmd ${OCI} 'up4:read-entities -i -g' 'tmp/Mid_check_interfaces.log'"/>
```
Comment (daniele-moro) [inline scenarios/pfcp-agent-failure.xml:31]: Remove

File: scenarios/pfcp-agent-failure.xml [hunk 4]
```diff
@@ -24,14 +27,25 @@
...
+
         <!-- Reboot SMF because it probably doesn't react well to failures -->
         <group name="AF-Restart-Smf" requires="~AF-Check-Flows-2">
-            <step name="AF-Kill-Smf"
-                  exec="mock-smf-cmd kill"/>
-            <step name="AF-Start-Smf" requires="^"
-                  exec="mock-smf-cmd start"/>
-            <step name="AF-Associate-Smf" requires="^" delay="1"
-                  exec="mock-smf-cmd associate"/>
+            <step name="AF-Kill-Smf" delay="1"
```
Comment (daniele-moro) [inline scenarios/pfcp-agent-failure.xml:35]: Do we need the delay or was only for tests?

File: scenarios/pfcp-agent-failure.xml [hunk 5]
```diff
@@ -24,14 +27,25 @@
...
+                  exec="${DOCKER_COMPOSE_CMD} restart mock-smf"/>
+            <step name="AF-Configure-Smf" requires="^"
+                  exec="${DOCKER_COMPOSE_CMD} exec -T mock-smf pfcpctl -c configure --n3-addr 140.0.0.1 --remote-peer pfcp-agent"/>
+            <step name="AF-Associate-Smf" requires="^"
+                  exec="${DOCKER_COMPOSE_CMD} exec -T mock-smf pfcpctl -c associate"/>
         </group>
+
+        <step name="AF-Post-Check-Interfaces" requires="AF-Restart-Smf"
+              exec="onos-cmd ${OCI} 'up4:read-entities -i -g' 'tmp/Post_check_interfaces.log'"/>
```
Comment (daniele-moro) [inline scenarios/pfcp-agent-failure.xml:44]: remove

File: scenarios/pfcp-agent-failure.xml [hunk 6]
```diff
@@ -24,14 +27,25 @@
...
+
+        <step name="AF-Check-Up4-N3-Interface" requires="AF-Restart-Smf"
+              exec="onos-cli-grep ${OCI} up4:read-interfaces 'UpfInterface(type=ACCESS, prefix=140.0.0.1/32'"/>
+        <step name="AF-Check-Up4-Uepool-Interface" requires="AF-Restart-Smf"
+              exec="onos-cli-grep ${OCI} up4:read-interfaces 'UpfInterface(type=CORE, prefix=17.0.0.0/16'"/>
```
Comment (daniele-moro) [inline scenarios/pfcp-agent-failure.xml:49]: This check can stay, and should help us quickly verify the restart was successful.

File: scenarios/setup.xml [hunk 7]
```diff
@@ -70,10 +70,10 @@
                   exec="onos-check-apps ${OC#} org.omecproject.up4 includes"
                   requires="Up4-App-Install-Local-Build"/>
             <step name="Up4-Set-Log-Debug-${#}"
-                  exec="onos ${OC#} log:set DEBUG org.omecproject.up4"
+                  exec="ssh -q -p ${OP#} -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null ${OC#} log:set DEBUG org.omecproject.up4"
                   requires="Onos-Check-Up4-App-${#}"/>
             <step name="Onos-Enable-Grpc-Log-${#}"
-                  exec="onos ${OC#} cfg set org.onosproject.grpc.ctl.GrpcChannelControllerImpl enableMessageLog true"
+                  exec="ssh -q -p ${OP#} -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null ${OC#} cfg set org.onosproject.grpc.ctl.GrpcChannelControllerImpl enableMessageLog true"
```
Comment (daniele-moro) [inline scenarios/setup.xml:76]: Can we move this to a different PR? Maybe adding a convenience script under `scenarios/bin` folder

File: app/Makefile [hunk 8]
```diff
@@ -54,7 +54,7 @@ build: _build_resources
 # ci-verify is a build profile specified in onos-dependencies's pom.xml
 # It's used to run javadoc validation and other checks that should not
 # run on local build, but during CI.
-build-ci: MVN_FLAGS := -B -q -Pci-verify -Pcoverage
+build-ci: MVN_FLAGS := -q -B -Pci-verify -Pcoverage
```
[the remark this answers is not in the document]
Comment (EmanueleGallone) [reply app/Makefile:57]: Nope, typo. Will fix

File: scenarios/pfcp-agent-failure.xml (hunk 4 above)
Comment (EmanueleGallone) [reply scenarios/pfcp-agent-failure.xml]: It was there for test. No need to delay that step.

File: scenarios/bin/onos-cmd (hunk 1 above)
Comment (EmanueleGallone) [reply scenarios/bin/onos-cmd]: Nope. Just for ease of development. Reverted to the original script

**Review by pierventre — APPROVED** [review]

File: app/app/src/main/java/org/omecproject/up4/impl/Up4DeviceManager.java [hunk 9]
```diff
@@ -1373,14 +1376,20 @@ private void checkFlowRuleStateAndReconcile() throws UpfProgrammableException {
...
                 Set<FlowRule> missingRules =
                     leaderRules.stream()
                         .filter(lr -> followerRules.stream().noneMatch(fr -> fr.equals(lr)))
                         .collect(Collectors.toSet());
                 ops.newStage();
-                missingRules.forEach(r -> ops.add(copyFlowRuleForDevice(r, deviceId)));
+                missingRules.forEach(r -> {
+                    log.trace("Adding {} to {}", copyFlowRuleForDevice(r, deviceId), deviceId);
+                    ops.add(copyFlowRuleForDevice(r, deviceId));
+                });
```
Comment (pierventre) [inline app/app/src/main/java/org/omecproject/up4/impl/Up4DeviceManager.java:1392]: do the same below for the meters

Comment (EmanueleGallone) [reply app/app/src/main/java/org/omecproject/up4/impl/Up4DeviceManager.java:1395]: Done. Please check

File: .env.stable [hunk 10]
```diff
@@ -2,8 +2,9 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
-PFCP_AGENT_IMAGE=omecproject/upf-epc-pfcpiface:master-9556adb
+PFCP_AGENT_IMAGE=omecproject/upf-epc-pfcpiface:master-f3a2c33
 ONOS_IMAGE=opennetworking/sdfabric-onos:master-2022-02-25
```
[the remark this answers is not in the document]
Comment (EmanueleGallone) [reply .env.stable]: Updated to master-2022-03-05

File: app/app/src/main/java/org/omecproject/up4/impl/Up4DeviceManager.java [hunk 11]
```diff
@@ -1440,7 +1452,10 @@ private void checkMeterStateAndReconcile() {
                         .forEach(m -> meterService.submit(m));
                 missingMeters.stream()
                         .map(m -> meterToMeterRequestForDevice(m, deviceId, true))
-                        .forEach(m -> meterService.submit(m));
+                        .forEach(m -> {
```
Comment (pierventre) [inline app/app/src/main/java/org/omecproject/up4/impl/Up4DeviceManager.java:1455]: you are missing the unexpected and the missing

Comment (EmanueleGallone) [reply app/app/src/main/java/org/omecproject/up4/impl/Up4DeviceManager.java:1461]: Now it should be ok

**Comment by EmanueleGallone** [comment]
retest this please

**Comment by pierventre** [comment]
feel free to merge it
